### PR TITLE
Reduce number of reductions in Appsignal.Instrumentation.instrument/3

### LINKED ISSUE #845: What can be done about performance? Adding Appsignal instrumentation increases execution and reductions by nearly 3x
If your bug is about how AppSignal works in your app specifically we recommend you contact us at support@appsignal.com with your bug report instead.

---

## Describe the bug
When adding Appsignal instrumentation to my app, via `@decorate transaction_event()` calls on functions, `:eprof` shows a rather large increase in reductions and execution time with Appsignal turned on.

Reduction difference 3.3x (2.4m wo AS | ~8m w AS)
Execution time difference 2.6x (0.66sec wo AS | 1.74sec w AS)

Here's `:eprof` with AppSignal, and underneath is without. I've tried to denote in the `with Appsignal` trace what the breakdown is on shared calls.

FYI the trace below is code that streams 1k lines from a file, and does some processing on each line of text(not a HUGE amount, but certainly a fair amount more than just some string split or trimming).

The legend for the character at the beginning of each line is as follows:

`M` = calls attributed solely to my app
`B` = calls that are split between both my app and Appsignal, i've tried  to denote on these how many calls my app makes when running the exact same function, without appsignal instrumenting. FYI the first number is my app without appsignal, the second number is with appsignal, and includes my apps calls too. So `658 vs   163158`, means appsignal is making 162,500 calls compared to my 658
`A` = calls attributed solely to Appsignal

```
...
M 'Elixir.String.Break':do_trim_trailing_l/1                             81250     0.47     8189  [      0.10]
M 'Elixir.Unicode.String.Segment':'-evaluate_rules/2-fun-0-'/3           92931     0.47     8241  [      0.09]
B erlang:whereis/1                                             658 vs   163158     0.48     8401  [      0.05]
M 'Elixir.Regex':'match?'/2                                             104483     0.50     8829  [      0.08]
A 'Elixir.Appsignal.Instrumentation':instrument/1                        40625     0.51     8892  [      0.22]
B ets:insert/2                                                    5 vs   40625     0.55     9663  [      0.24]
A 'Elixir.Appsignal.Config':'do_valid?'/1                                81250     0.57    10053  [      0.12]
B 'Elixir.String.Chars':'impl_for!'/1                         110 vs    121985     0.59    10341  [      0.08]
M 'Elixir.Unicode.String.Segment':evaluate_rule/2                        92931     0.60    10473  [      0.11]
A 'Elixir.Appsignal.Config':'valid?'/0                                   81250     0.60    10499  [      0.13]
A 'Elixir.Appsignal.Tracer':create_span/3                                40625     0.61    10614  [      0.26]
A 'Elixir.Appsignal.Tracer':lookup/1                                     81250     0.61    10744  [      0.13]
B 'Elixir.String.Chars':impl_for/1                             110 vs   121985     0.62    10835  [      0.09]
B 'Elixir.Keyword':get/3                                       44846 vs 126096     0.62    10924  [      0.09]
A 'Elixir.Appsignal.Nif':'_set_span_name'/2                              40625     0.67    11776  [      0.29]
M re:internal_run/4                                                      16879     0.68    11829  [      0.70]
B 'Elixir.String.Chars':to_string/1                             110 vs  121985     0.69    12046  [      0.10]
B application:get_env/3                                         20 vs   162500     0.78    13593  [      0.08]
A 'Elixir.Appsignal.Tracer':'running?'/0                                162500     0.78    13605  [      0.08]
B 'Elixir.Application':get_env/3                                0 vs    162500     0.82    14366  [      0.09]
B 'Elixir.Process':whereis/1                                   358 vs   162858     0.85    14790  [      0.09]
A 'Elixir.Appsignal.Nif':'_set_span_attribute_string'/3                  40625     0.87    15129  [      0.37]
M 'Elixir.Regex':version/0                                              115908     0.91    15875  [      0.14]
M 'Elixir.Regex':safe_run/3                                             115908     1.00    17513  [      0.15]
B erlang:send/2                                                   6 vs   40625     1.65    28820  [      0.71]
B application_controller:get_env/2                                5 vs  162500     2.60    45500  [      0.28]
A 'Elixir.Appsignal.Nif':'_create_child_span'/1                          20084     4.68    81913  [      4.08]
M re:run/3                                                              115908     5.11    89312  [      0.77]
A 'Elixir.Appsignal.Nif':'_create_root_span'/1                           20541     5.25    91802  [      4.47]
B ets:lookup/2                                                  200 vs  243950     5.46    95554  [      0.39]
A 'Elixir.Appsignal.Nif':'_close_span'/1                                 40625    10.16   177689  [      4.37]
M gen:do_call/4                                                            658    27.02   472452  [    718.01]
-------------------------------------------------------------------  -------  -------  -------  [----------]
Total:                                                               7949086  100.00%  1748640  [      0.22]
```

Below is the same trace as above, except with appsignal instrumentation commented out and recompiled before running. Both of these traces have been truncated, mostly because it gets very long, and the stuff above this point has quickly diminishing impact on the overall execution time.

```
...
unicode:characters_to_binary/2                                          4800     0.31    2067  [      0.43]
'Elixir.String':replace_guarded/4                                       4800     0.31    2080  [      0.43]
re:process_parameters/6                                                34675     0.33    2226  [      0.06]
lists:keyfind/3                                                        45148     0.35    2328  [      0.05]
'Elixir.Unicode.String.Break':split/3                                   7286     0.35    2371  [      0.33]
'Elixir.Regex':split/3                                                 11226     0.37    2480  [      0.22]
'Elixir.Enum':uniq_list/3                                              35880     0.37    2489  [      0.07]
'Elixir.MapSet':'member?'/2                                            24422     0.38    2524  [      0.10]
'Elixir.String.Unicode':downcase/3                                     38225     0.42    2811  [      0.07]
maps:fold/3                                                            22014     0.44    2958  [      0.13]
maps:next/1                                                            47636     0.45    3012  [      0.06]
binary:split/3                                                          4900     0.51    3379  [      0.69]
maps:fold_1/3                                                          44028     0.56    3758  [      0.09]
'Elixir.Keyword':get/3                                                 44846     0.60    4019  [      0.09]
'Elixir.Kernel':struct/3                                               44028     0.61    4063  [      0.09]
'Elixir.MyApp':do_xxx/2                                                 8895     0.72    4783  [      0.54]
'Elixir.Map':get/3                                                     38412     0.81    5384  [      0.14]
re:version/0                                                          115908     0.93    6234  [      0.05]
erlang:system_info/1                                                  115908     0.95    6369  [      0.05]
'Elixir.Enum':'-map/2-lists^map/1-0-'/2                                50440     1.09    7255  [      0.14]
'Elixir.Enumerable.List':reduce/3                                     100217     1.24    8278  [      0.08]
'Elixir.Unicode.String.Segment':'-evaluate_rules/2-fun-0-'/3           92931     1.25    8335  [      0.09]
'Elixir.Regex':'match?'/2                                             104483     1.30    8697  [      0.08]
ets:lookup/2                                                             200     1.42    9488  [     47.44]
'Elixir.Unicode.String.Segment':evaluate_rule/2                        92931     1.59   10623  [      0.11]
re:internal_run/4                                                      16879     1.89   12639  [      0.75]
'Elixir.Regex':version/0                                              115908     2.30   15392  [      0.13]
'Elixir.Regex':safe_run/3                                             115908     2.59   17319  [      0.15]
re:run/3                                                              115908    14.24   95200  [      0.82]
gen:do_call/4                                                            658    48.81  326207  [    495.76]
-------------------------------------------------------------------  -------  -------  ------  [----------]
Total:                                                               2383271  100.00%  668374  [      0.28]
```
Is there maybe something I'm doing wrong that would cause such a drastic increase? Currently all i'm doing is as follows:

```
defmodule SomeModule do
  use Appsignal.Instrumentation.Decorators

  @decorate transaction_event()
  def some_function(), do: :ok
end
```
Thanks!

One correction, I use the term `reduction` incorrectly above, what I really mean is function calls. `658` is the number of times `gen:do_call/4` is called, _not_ the number of reductions that `gen:do_call/4` consumed.

### PR DESCRIPTION
**Pull request by jeffkreeftmeijer**
As reported in https://github.com/appsignal/appsignal-elixir/issues/845, the instrumentation helper did some unnecessary reductions. This pull request is the result of looking into the reductions report and removing some defensive programming we didn't really need, _without_ changing any functionality.

## eprof

Elixir's `profile.eprof` mix task evaluates a piece of code with the [eprof](https://www.erlang.org/doc/man/eprof.html) profiling tool enabled. After running the code, it outputs a time profile report, which shoes the number of calls to each function, as well as the amount of time spent in each.

To look into the `Appsignal.Instrumentation.instrument/3` function, we evaluated the following code snippet through the eprof task:

```shell
mix profile.eprof -e "(1..1_000) |> Enum.each(fn _ -> Appsignal.Instrumentation.instrument(\"name\", \"category\", fn -> :ok end) end)"
```

For convenience, this is the executed code, which calls the `instrument/3` function a thousand times to get rid of any outliers.

```elixir
(1..1_000)
|> Enum.each(fn _ ->
  Appsignal.Instrumentation.instrument(\"name\", \"category\", fn -> :ok end)
end)
```

<a id="org69b9b93"></a>

## The report

Running the report for the first time produced the following result:

```
Profile results of #PID<0.315.0>
#                                                        CALLS     %  TIME µS/CALL
Total                                                    11701 100.0 53586    0.46
anonymous fn/0 in :elixir_compiler_2.__FILE__/1              1  0.00     1    1.00
:erlang.apply/2                                              1  0.01     3    3.00
Enum.each/2                                                  1  0.01     7    7.00
anonymous fn/0 in :elixir_compiler_2.__FILE__/1           1000  0.22   117    0.12
Appsignal.Nif.close_span/1                                1000  0.22   119    0.12
Keyword.get/2                                             1000  0.22   119    0.12
Appsignal.Nif.set_span_attribute_string/3                 1000  0.22   120    0.12
List.last/1                                               1000  0.23   122    0.12
Appsignal.Nif.create_root_span/1                          1000  0.23   125    0.13
Appsignal.Nif.set_span_name/2                             1000  0.24   126    0.13
GenServer.cast_msg/1                                      1000  0.25   132    0.13
Appsignal.Instrumentation.instrument/3                    1000  0.25   135    0.14
Appsignal.Tracer.create_span/2                            1000  0.25   136    0.14
Appsignal.Tracer.current_span/0                           1000  0.28   152    0.15
anonymous fn/1 in :elixir_compiler_2.__FILE__/1           1000  0.28   152    0.15
Appsignal.Monitor.add/0                                   1000  0.29   156    0.16
Appsignal.Tracer.current_span/1                           1000  0.43   233    0.23
Application.get_env/2                                     2000  0.44   234    0.12
Access.get/2                                              2000  0.44   237    0.12
String.trim_trailing/1                                    2000  0.45   241    0.12
String.trim_leading/1                                     2000  0.46   244    0.12
String.Break.trim_leading/1                               2000  0.48   258    0.13
Appsignal.Tracer.deregister/1                             1000  0.51   271    0.27
String.Break.trim_trailing/1                              2000  0.52   276    0.14
Enum.reduce_range/5                                       1001  0.53   285    0.28
GenServer.do_send/2                                       1000  0.54   287    0.29
:lists.keyfind/3                                          2000  0.54   288    0.14
Appsignal.Span.close/1                                    1000  0.54   288    0.29
GenServer.cast/2                                          1000  0.54   290    0.29
Appsignal.Config.do_active?/1                             2000  0.55   296    0.15
anonymous fn/3 in Enum.each/2                             1000  0.56   298    0.30
Appsignal.Instrumentation.call_with_optional_argument/2   2000  0.56   300    0.15
Appsignal.Span.set_attribute/3                            1000  0.56   300    0.30
Access.get/3                                              2000  0.58   310    0.16
Appsignal.Nif._set_span_name/2                            1000  0.61   327    0.33
Appsignal.Config.empty?/1                                 2000  0.62   330    0.17
Appsignal.Tracer.current/1                                2000  0.67   358    0.18
String.Break.do_trim_leading/1                            2000  0.69   371    0.19
Appsignal.Tracer.ignored?/1                               2000  0.73   393    0.20
Appsignal.Tracer.register/1                               1000  0.74   394    0.39
anonymous fn/4 in Appsignal.Instrumentation.instrument/3  1000  0.74   396    0.40
:ets.delete_object/2                                      1003  0.77   411    0.41
Appsignal.Span.set_name/2                                 1000  0.78   419    0.42
Appsignal.Span.create_root/2                              1000  0.80   431    0.43
:ets.insert/2                                             1003  0.84   449    0.45
Application.maybe_warn_on_app_env_key/2                   4000  0.88   472    0.12
Keyword.get/3                                             2000  0.90   484    0.24
:application.get_env/2                                    4000  0.90   484    0.12
Appsignal.Tracer.close_span/1                             1000  0.95   510    0.51
Appsignal.Config.active?/0                                2000  1.00   535    0.27
String.trim/1                                             2000  1.01   540    0.27
String.Break.do_trim_trailing_l/1                         2000  1.02   544    0.27
:erlang.send/2                                            1000  1.13   603    0.60
String.Break.trim_trailing/2                              2000  1.24   662    0.33
Appsignal.Instrumentation.instrument/1                    1000  1.28   688    0.69
Appsignal.Config.do_valid?/1                              2000  1.30   694    0.35
Appsignal.Tracer.lookup/1                                 2000  1.32   707    0.35
Appsignal.Config.valid?/0                                 2000  1.40   749    0.37
:erlang.whereis/1                                         4000  1.40   751    0.19
Appsignal.Tracer.create_span/3                            1000  1.50   802    0.80
Appsignal.Nif._set_span_attribute_string/3                1000  1.68   901    0.90
:application.get_env/3                                    4000  1.77   949    0.24
Process.whereis/1                                         4000  1.82   977    0.24
Application.get_env/3                                     4000  1.85   994    0.25
:application_controller.get_env/2                         4000  1.91  1026    0.26
Appsignal.Tracer.running?/0                               4000  2.13  1143    0.29
:ets.lookup/2                                             6000  7.62  4083    0.68
Appsignal.Nif._create_root_span/1                         1000 13.82  7404    7.40
Appsignal.Nif._close_span/1                               1000 29.76 15947   15.95
```

The report shows all function calls, ordered by thir execution times. The rightmost column shows the time per call in microseconds, showing that the `Appsignal.Nif._close_span/1` and `Appsignal.Nif._create_root_span/1` functions take the most time, at roughly 16 and 7 microseconds per call. Both functions are called 1000 times, which makes sense because that's the amount of spans that are started and stopped. Although there might be some performance gain to be had here, we're skipping these function calls, as they're called a reasonable amount of times.

<a id="org9fade91"></a>

## `:ets.lookup/2`

One place above these is `:ets.lookup/2`, which only takes 0.68 microseconds per call, but it's called 6000 times, or $6n$. Although AppSignal uses ets internally to store the currently open spans, some of these calls actually originated from checking the configuration.

Internally, the integration had a check to see if it's active before calling `Span.set_name/2`, which is the function used to set the name for a span. However, this call was unneeded, as calling this function with a `nil` value instead of a span is a no-op. When the integration isn't active, nils are passed around instead of spans, so this extra check can be safely removed.

This reduced the number calls by $2n$, leaving $4n$:

```
:ets.lookup/2                                             4000  5.06  2458    0.61
Appsignal.Nif._create_root_span/1                         1000 16.30  7922    7.92
Appsignal.Nif._close_span/1                               1000 32.91 16001   16.00
```

Another :ets lookup was caused by a duplicate call to `Application.get_env/1`. Since the application environment is also stored in ets, each `get_env/1` call also does a lookup.

In AppSignal, the environment was once in the call to `active?/0`, which checks in the config is configured to be, and once in `valid?/0`, which checks if the API key is valid. If either of these is false, the integration is disabled. This patch includes a fix that passes the environment that was already fetched by calling `active?/0` to `valid?/1`, saving another reduction:

```
:erlang.whereis/1                                         4000  1.88   757    0.19
Appsignal.Nif._set_span_attribute_string/3                1000  1.90   766    0.77
Appsignal.Tracer.create_span/3                            1000  2.02   813    0.81
Process.whereis/1                                         4000  2.17   875    0.22
Appsignal.Tracer.running?/0                               4000  2.30   926    0.23
:ets.lookup/2                                             3001  3.91  1576    0.53
Appsignal.Nif._create_root_span/1                         1000 17.12  6903    6.90
Appsignal.Nif._close_span/1                               1000 33.97 13696   13.70
```

Now, the number of calls to `:ets.lookup/2` is $3n$. One call to try and find the parent, another call to see if the current PID is ignored, and a last call to check if the integration is currently active.

It might be possible to remove the check to see if the integration is active that happens before starting every span. However, this patch only consisting of performance enhancements, changes in functionality aren't included.

<a id="orgf64075e"></a>

## `Appsignal.Tracer.running?/0`

A line above the `:ets.lookup/2` function is `Appsignal.Tracer.running?/0`, a function that's called to make sure ets is running before executing lookups, inserts and deletions. Because ets produces an `ArgumentError` when it's not active, the integration called this check to make sure the registry was running every time an evaluation was done on the table.

This patch removes those checks and adds error handling around each function that calls out to ets, which removes `Appsignal.Tracer.running?/0` completely.

Closes #845.

### CODE REVIEW

**Comment by unflxw** [comment]
This is good stuff! Would it be possible to add some sort of diff, between the report before and after the change?

**Comment by jeffkreeftmeijer** [comment]
> This is good stuff! Would it be possible to add some sort of diff, between the report before and after the change?

Here's a diff of the report before (on the `main` branch) and after (on the `eprof` branch): https://gist.github.com/jeffkreeftmeijer/f7df1d6142d1f3c0bab6539b4dd0946a/revisions. I couldn't get it to word-diff, but maybe someone could, given those two report revisions.

**Comment by unflxw** [comment]
Thank you! I'm not sure what's up with the totals in the diff, but regardless, you can see on the diff how the changes "trickle down" the call stack -- many function calls are reduced, not just the ones explicitly mentioned in the write up above. I think it reinforces the point.

File: lib/appsignal/tracer.ex [hunk 1]
```diff
@@ -49,37 +49,33 @@ defmodule Appsignal.Tracer do
...
   Finds the span in the registry table.
   """
-  @spec lookup(pid()) :: list()
+  @spec lookup(pid()) :: list() | nil
   def lookup(pid) do
-    if running?(), do: :ets.lookup(@table, pid)
+    try do
+      :ets.lookup(@table, pid)
+    rescue
+      ArgumentError -> []
```
Comment (unflxw) [inline lib/appsignal/tracer.ex:77]: Going by the type annotation, is the rescue class meant to return `nil`, not `[]`?

**Review by unflxw — APPROVED** [review]

**Review by tombruijn — APPROVED** [review]

**Pull request merged**